applications/spectrum_analyzer/spectrum_analyzer.c

```diff
@@ -0,0 +1,475 @@
+
+        if(event_status == osOK) {
+
+            // FURI_LOG_D("Spectrum", "Main Loop - event: %u", event.type);
+
+            uint8_t vstep = VERTICAL_SHORT_STEP;
+            uint8_t hstep;
+            switch (width) {
+            case NARROW:
+                hstep = NARROW_STEP;
+                break;
+            case ULTRAWIDE:
+                hstep = ULTRAWIDE_STEP;
+                break;
+            default:
+                hstep = WIDE_STEP;
+                break;
+            }
+
+            // press events
+            if(event.type == EventTypeKey) {
+                if(event.input.type == InputTypePress) {  
+                    switch(event.input.key) {
+                    case InputKeyUp: 
+                        vscroll = MAX(vscroll - vstep, MIN_VSCROLL);
+                        FURI_LOG_D("Spectrum", "Vscroll: %u", vscroll);
+                        break; 
+                    case InputKeyDown: 
+                        vscroll = MIN(vscroll + vstep, MAX_VSCROLL);
+                        FURI_LOG_D("Spectrum", "Vscroll: %u", vscroll);
+                        break; 
+                    case InputKeyRight: 
+                        user_freq += hstep;
+                        FURI_LOG_D("Spectrum", "user_freq: %lu", user_freq);
+                        update_values_flag = true;
+                        break; 
+                    case InputKeyLeft:  
+                        user_freq -= hstep;
+                        FURI_LOG_D("Spectrum", "user_freq: %lu", user_freq);
+                        update_values_flag = true;
+                        break; 
+                    case InputKeyOk: 
+                        {
+                            switch (width) {
+                                case WIDE:
+                                    width = NARROW;
+                                    break;
+                                case NARROW:
+                                    width = ULTRAWIDE;
+                                    break;
+                                case ULTRAWIDE:
+                                default:
+                                    width = WIDE;
+                            }
+                        }
+                        update_values_flag = true;
+                        FURI_LOG_D("Spectrum", "Width: %u", width);
+                        break;
+                    case InputKeyBack: 
+                        processing = false;
+                        break;
+                    }
+                }
+            } 
+        } else {
+            FURI_LOG_T("spectrum", "osMessageQueue: event timeout");
+            // event timeout
+        }
+        view_port_update(view_port);
+        release_mutex(&model_mutex, model);
+    }
+    
+    view_port_enabled_set(view_port, false);
+    gui_remove_view_port(gui, view_port);
+    furi_record_close("gui");
+    view_port_free(view_port);
+    
+    osThreadTerminate(analyzer_thread);
```

Comment:
It's prohibited to use this function. It's extremely unsafe: terminating thread is going to leave system in inconsistent state that is impossible to predict and resolve.

Please use furi_thread. Also please check new version of music_player, i've removed this anti-pattern.

Reply:
using furi_thread